truffle/src/com.oracle.truffle.api.strings/src/com/oracle/truffle/api/strings/TSTaintNodes.java

```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    @GenerateUncached
+    public abstract static class IsArrayTaintedNode extends Node {
+
+        IsArrayTaintedNode() { }
+
+        /**
+         * Checks whether the given array is tainted.
+         * An array is tainted if it is not {@code null} and has at least
+         * one entry which is also not {@code null}.
+         */
+        public abstract boolean execute(Object[] taint);
+
+        @Specialization(guards = "taint != null")
+        static boolean isTaintedNonNull(Object[] taint,
+                                 @Cached IsSubArrayTaintedNode isSubArrayTaintedNode) {
+            return isSubArrayTaintedNode.execute(taint, 0, taint.length);
+        }
+
+        @Specialization(guards = "taint == null")
+        static boolean isTaintedNull(Object[] taint) {
+            return false;
+        }
+
+        static IsArrayTaintedNode getUncached() {
+            return TSTaintNodesFactory.IsArrayTaintedNodeGen.getUncached();
+        }
+    }
+
+    @GenerateUncached
+    public abstract static class IsTaintedNode extends Node {
+
+        IsTaintedNode() { }
+
+        /**
+         * Checks if the given {@link AbstractTruffleString} is tainted.
+         * A {@link AbstractTruffleString} is tainted iff {@link AbstractTruffleString#taint()} is not {@code null}
+         * and contains at least one element, which is also not {@code null}.
+         */
+        public abstract boolean execute(AbstractTruffleString a);
+
+        @SuppressWarnings("unused")
+        @Specialization
+        static boolean isTainted(AbstractTruffleString a,
+                                 @Cached IsArrayTaintedNode isArrayTaintedNode) {
+            return isArrayTaintedNode.execute(a.taint());
+        }
+
+        static IsTaintedNode getUncached() {
+            return TSTaintNodesFactory.IsTaintedNodeGen.getUncached();
+        }
+    }
+
+    @GenerateUncached
+    public abstract static class AddTaintNode extends Node {
+
+        AddTaintNode() { }
+
+        /**
+         * Adds taint to the given {@link AbstractTruffleString}.
+         */
+        public abstract AbstractTruffleString execute(AbstractTruffleString a, Object taint);
+
+        @SuppressWarnings("unused")
+        @Specialization
+        static AbstractTruffleString addTaint(TruffleString a, Object taint) {
+            final Object[] taintArr = new Object[a.codePointLength()];
+            Arrays.fill(taintArr, taint);
+            return TruffleString.createFromArray(
+                    a.data(),
+                    a.offset(),
+                    a.length(),
+                    a.stride(),
+                    a.encoding(),
+                    a.codePointLength(),
+                    a.codeRange(),
+                    true,
+                    taintArr
+            );
+        }
+
+        // TODO should adding taint to a MutableTruffleString produce a new object?
```

Comment:
How mutable should `MutableTruffleString`s be? Should adding / removing taint create a new object or change the existing one?

Reply:
`addTaint` now creates always a new object.
Also `addTaint` and `removeTaint` throw now an `UnsupportedOperationException` if called with `MutableTruffleString`